File: C/wayland/wayland-01-list_interfaces.c
/*
* wayland-01-list_interfaces.c
* Copyright (C) 2025  Manuel Bachmann <tarnyko.tarnyko.net>
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*  Prerequisites:
 - Debian/Ubuntu: $ sudo apt install libwayland-dev
 - Fedora/RHEL:   $ sudo dnf install wayland-devel

    Compile with:
 $ gcc ... `pkg-config --cflags --libs wayland-client`
*/

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <wayland-client.h>


typedef enum {
    E_UNKNOWN = 0, E_WESTON = 1, E_GNOME = 2, E_KDE = 3, E_WLROOTS = 4
} CompositorId;

typedef struct {
    CompositorId id;
} InterfaceInfo;


void wl_interface_available(void*, struct wl_registry*, uint32_t, const char*, uint32_t);
void wl_interface_removed(void*, struct wl_registry*, uint32_t);

static const struct wl_registry_listener wl_registry_listener = {
    wl_interface_available,
    wl_interface_removed
};


int main(int argc, char* argv[])
{
    struct wl_display* display = wl_display_connect(NULL);
    if (!display) {
        fprintf(stderr, "No Wayland compositor found! Do you have a '$XDG_RUNTIME_DIR/wayland-0' socket?\n"
                        "If not, start it, and set environment variables:\n"
                        "$ export XDG_RUNTIME_DIR=/run/user/$UID\n"
                        "$ export WAYLAND_DISPLAY=wayland-0\n\n");
        return EXIT_FAILURE;
    }

    struct wl_registry* registry = wl_display_get_registry(display);
    assert(registry);

    // listen for asynchronous callbacks with an '_info' struct attached
    InterfaceInfo _info = {0};
    wl_registry_add_listener(registry, &wl_registry_listener, &_info);

    // wait for a compositor roundtrip, so all callbacks are fired...
    wl_display_roundtrip(display);

    // ... and '_info' has now been filled by 'wl_interface_available()'
    printf("\nCompositor is: ");
    switch (_info.id)
    {
        case E_WESTON : printf("Weston.\n\n");     break;
        case E_GNOME  : printf("GNOME.\n\n");      break;
        case E_KDE    : printf("KDE Plasma.\n\n"); break;
        case E_WLROOTS: printf("wlroots.\n\n");    break;
        default       : printf("Unknown...\n\n");
    }

    // clean
    wl_registry_destroy(registry);
    wl_display_flush(display);
    wl_display_disconnect(display);

    return EXIT_SUCCESS;
}


// WL_REGISTRY CALLBACKS

void wl_interface_available(void* data, struct wl_registry* registry, uint32_t serial, const char* name, uint32_t version)
{
    printf("Interface available: name:'%s' - version:'%d'. ", name, version);

    InterfaceInfo* _info = data;

         if (!strcmp(name, "wl_shm"))                  { printf("\t\t\t [Software rendering]");                               }
    else if (!strcmp(name, "wl_seat"))                 { printf("\t\t\t [Input devices (keyboard, mouse, touch)]");           }
    else if (!strcmp(name, "wl_output"))               { printf("\t\t\t [Output devices (screens)]");                         }
    else if (!strcmp(name, "wl_data_device_manager"))  { printf("    \t [Clibpoard (copy-paste, drag-drop)]");                }
    else if (!strcmp(name, "wp_viewporter"))           { printf("  \t\t [Surface scaling]");                                  }
    else if (!strcmp(name, "wp_presentation"))         { printf("  \t\t [Precise video synchronization]");                    }
    else if (strstr(name,  "wp_idle_inhibit_manager")) { printf("\b\b\t [Screensaver inhibiter]");                            }
    else if (strstr(name,  "wp_text_input_manager"))   { printf("\b\b\t [Virtual keyboard]");                                 }
    else if (strstr(name,  "wp_pointer_constraints"))  { printf("\b\b\t [Pointer lock]");                                     }
    else if (strstr(name,  "wp_linux_dmabuf"))         { printf("    \t [DRM Kernel GPU channel]");                           }
    else if (!strcmp(name, "wl_drm"))                  { printf("\t\t\t [DRM kernel GPU channel -deprecated]");               }
    else if (!strcmp(name, "wl_shell"))                { printf("\t\t\t [Standard window manager -deprecated]");              }
    else if (!strcmp(name, "xdg_wm_base"))             { printf("\t\t\t [Standard window manager]");                          }
    else if (strstr(name,  "xdg_shell"))               { printf("  \t\t [Standard window manager -unstable]");                }
    else if (strstr(name,  "gtk_shell"))               { printf("  \t\t [GNOME window manager]");      _info->id = E_GNOME;   }
    else if (strstr(name,  "plasma_shell"))            { printf("  \t\t [KDE Plasma window manager]"); _info->id = E_KDE;     }
    else if (strstr(name,  "wlr_layer_shell"))         { printf("    \t [wlroots window manager]");    _info->id = E_WLROOTS; }
    else if (strstr(name,  "weston"))                  {                                               _info->id = E_WESTON;  }
    else if (!strcmp(name, "wl_subcompositor"))        { printf("  \t\t [Sub-surfaces]");                                     }
    else if (!strcmp(name, "wl_compositor"))           { printf("  \t\t [Compositor]");                                       }

    putchar('\n');
}

void wl_interface_removed(void* data, struct wl_registry* registry, uint32_t serial)
{ }

File: Rust/wayland/src/bin/wayland-01-list_interfaces.rs
use std::error::Error;

use wayland_client::{
    Connection, EventQueue, QueueHandle, Dispatch,
    protocol::wl_registry,
};


const BS: char = 8u8 as char;

enum CompositorId { Unknown, Weston, GNOME, KDE, WlRoots }

struct InterfaceInfo { id: CompositorId }


fn main() -> Result<(), Box<dyn Error>>
{
    let Ok(connection) = Connection::connect_to_env() else {
        return Err("No Wayland compositor found! Do you have a '$XDG_RUNTIME_DIR/wayland-0' socket?\n
                    If not, start it, and set environment variables:\n
                    $ export XDG_RUNTIME_DIR=/run/user/$UID\n
                    $ export WAYLAND_DISPLAY=wayland-0\n\n".into());
    };

    let display = connection.display();

    // listen for asynchronous callbacks with an 'InterfaceInfo' struct attached
    let mut queue: EventQueue<InterfaceInfo> = connection.new_event_queue();
    let queue_h = queue.handle();
    display.get_registry(&queue_h, ());

    // wait for a compositor roundtrip, so all callbacks are fired...
    let mut info = InterfaceInfo { id: CompositorId::Unknown };
    queue.roundtrip(&mut info).unwrap();

    // .. and 'info' has now been filled by 'Dispatch<> -> event()'
    print!("\nCompositor is: ");
    match info.id
    {
        CompositorId::Weston  => println!("Weston.\n"),
        CompositorId::GNOME   => println!("GNOME.\n"),
        CompositorId::KDE     => println!("KDE Plasma.\n"),
        CompositorId::WlRoots => println!("wlroots.\n"),
        _                     => println!("Unknown...\n")
    }

    Ok(())
}


// WL_REGISTRY_CALLBACKS

impl Dispatch<wl_registry::WlRegistry, ()> for InterfaceInfo {
  fn event(this: &mut Self,
           _: &wl_registry::WlRegistry, event: wl_registry::Event,
           _: &(), _: &Connection, _: &QueueHandle<InterfaceInfo>)
  {
      if let wl_registry::Event::Global { interface, version, .. } = event {
          print!("Interface available: name:'{}' - version:'{}'.", interface, version);

          match interface {
            _ if interface == "wl_shm"                            => print!(    "\t\t\t [Software rendering]"),
            _ if interface == "wl_seat"                           => print!(    "\t\t\t [Input devices (keyboard, mouse, touch)]"),
            _ if interface == "wl_output"                         => print!(    "\t\t\t [Output devices (screens)]"),
            _ if interface == "wl_data_device_manager"            => print!(    "    \t [Clipboard (copy-paste, drag-drop)]"),
            _ if interface == "wp_viewporter"                     => print!(    "  \t\t [Surface scaling]"),
            _ if interface == "wp_presentation"                   => print!(    "  \t\t [Precise video synchronization -old]"),
            _ if interface.contains("wp_fifo_manager")            => print!(  "   {0}\t [Precise video synchronization]", BS),
            _ if interface.contains("wp_tearing_control_manager") => print!(  "{0}{0}\t [Tearing control]", BS),
            _ if interface.contains("wp_idle_inhibit_manager")    => print!(  "{0}{0}\t [Screensaver inhibiter]", BS),
            _ if interface.contains("wp_tablet_manager")          => print!(    "    \t [Graphics tablet input]"),
            _ if interface.contains("wp_text_input_manager")      => print!(  "{0}{0}\t [Virtual keyboard]", BS),
            _ if interface.contains("wp_pointer_gestures")        => print!(  "{0}{0}\t [Touchpad gestures]", BS),
            _ if interface.contains("wp_pointer_constraints")     => print!(  "{0}{0}\t [Pointer lock]", BS),
            _ if interface.contains("wp_relative_pointer_manager")=> print!(" {0}{0}{0} [Pointer lock movement]", BS),
            _ if interface.contains("wp_linux_dmabuf")            => print!(   "     \t [DRM Kernel GPU channel]"),
            _ if interface == "wl_drm"                            => print!(    "\t\t\t [DRM Kernel GPU channel -deprecated]"),
            _ if interface == "wl_shell"                          => print!(    "\t\t\t [Standard window manager -deprecated]"),
            _ if interface == "xdg_wm_base"                       => print!(    "\t\t\t [Standard window manager]"),
            _ if interface.contains("xdg_shell")                  => print!(    "  \t\t [Standard window manager -unstable]"),
            _ if interface.contains("gtk_shell")                  => { print!(  "  \t\t [GNOME window manager]");
                                                                       this.id = CompositorId::GNOME }
            _ if interface.contains("plasma_shell")               => { print!(  "  \t\t [KDE Plasma window manager]");
                                                                       this.id = CompositorId::KDE },
            _ if interface.contains("wlr_layer_shell")            => { print!(  "  \t\t [wlroots window manager]");
                                                                       this.id = CompositorId::WlRoots },
            _ if interface.contains("weston")                     => { this.id = CompositorId::Weston },
            _ if interface.contains("zxdg_exporter_v2")           => print!(   "     \t [Foreign client surface export]"),
            _ if interface.contains("zxdg_importer_v2")           => print!(   "     \t [Foreign client surface import]"),
            _ if interface.contains("zxdg_exporter_v1")           => print!(   "     \t [Foreign client surface export -old]"),
            _ if interface.contains("zxdg_importer_v1")           => print!(   "     \t [Foreign client surface import -old]"),
            _ if interface.contains("xdg_activation")             => print!(   "     \t [Window focus switcher]"),
            _ if interface.contains("xdg_output_manager")         => print!(  "{0}{0}\t [Output devices (screens) dimensions]", BS),
            _ if interface.contains("xdg_decoration_manager")     => print!(  "{0}{0}\t [Server-side window decorations]", BS),
            _ if interface.contains("ext_idle_notifier")          => print!(   "     \t [Client-side idle notifier]"),
            _ if interface.contains("ext_workspace_manager")      => print!(  "{0}{0}\t [Virtual desktops]", BS),
            _ if interface == "wl_subcompositor"                  => print!(    "  \t\t [Sub-surfaces]"),
            _ if interface == "wl_compositor"                     => print!(    "  \t\t [Compositor]"),
            _ => ()
          }

          println!("");
      }
    }
}

